[GVN] Improve processBlock for instruction erasure

madhur13490 opened the pull request:
This patch deletes the instructions right away by using the approapriate iterators. Thus avoids collecting the instructions in a vector and then doing the erasure.

madhur13490 committed:
[GVN] Improve processBlock for instruction erasure

This patch deletes the instructions right away by using
the approapriate iterators. Thus avoids collecting the instructions
in a vector and then doing the erasure.

madhur13490 commented:
All tests pass but not sure how to construct a test for this patch. Suggestions welcome!

nikic reviewed (commented):
Could you provide more details on the motivation for this change? Is it just to resolve the FIXME or something else?

It looks like the stack-layout-remark.c clang test crashes.

madhur13490 commented:
> Could you provide more details on the motivation for this change? Is it just to resolve the FIXME or something else?
yes, primarily FIXME. Nothing else as such which motivates. The change is an optimization in the process of handling instructions by avoiding vector population just for the erasure.

> It looks like the stack-layout-remark.c clang test crashes.
Fixed in the current revision. Thanks!

madhur13490 commented:
@nikic Gentle ping!

madhur13490 commented:
Ping to reviewers! @antoniofrighetto can you please have a look?

antoniofrighetto commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -2820,41 +2817,12 @@ bool GVNPass::processBlock(BasicBlock *BB) {
-
-  for (BasicBlock::iterator BI = BB->begin(), BE = BB->end();
-       BI != BE;) {
+  for (BasicBlock::iterator BI = BB->begin(), BE = BB->end(); BI != BE;) {
+    Instruction *Inst = &*BI++;
```
```suggestion
  for (Instruction &Inst : make_early_inc_range(*BB)) {
```

antoniofrighetto commented on llvm/include/llvm/Transforms/Scalar/GVN.h:
```diff
@@ -25,6 +25,8 @@
 #include "llvm/IR/ValueHandle.h"
 #include "llvm/Support/Allocator.h"
 #include "llvm/Support/Compiler.h"
+#include "llvm/Transforms/Utils/AssumeBundleBuilder.h"
```
Can remove this from the C++ file?

madhur13490 replied:
Done

madhur13490 replied:
Done

nikic replied:
Can drop these includes now.

madhur13490 replied:
Can't remove. salvage* is provided by `AssumeBundleBuilder`.

nikic replied:
Yes, but it's no longer necessary in the header. You can drop the includes in the header and restore the AssumeBundleBuilder.h include in the cpp file.

madhur13490 replied:
> You can drop the includes in the header

Do you mean just the `AssumeBundleBuilder`?

nikic replied:
I mean the two you added.

madhur13490 committed:
Address review comments

madhur13490 commented:
@antoniofrighetto @nikic Addressed all review comments. The thing about guarding assert under `V != nullptr` in `verifyRemoved()` was added for some test but I don't remember which one. Running `ninja check-all` today leads to no new failures, thus I have removed that change. Probably the issue is fixed somewhere else.

madhur13490 commented:
Gentle ping @antoniofrighetto @nikic !

nikic commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -877,6 +876,12 @@ void GVNPass::printPipeline(
   OS << '>';
 }
 
+void GVNPass::doInstructionDeletion(Instruction *I) {
```
From the name, it's not really obvious what the difference between removeInstruction and doInstructionDeletion is. Maybe call this salvageAndRemoveInstruction?

madhur13490 replied:
Sure, sounds good.

nikic commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -2206,13 +2210,13 @@ bool GVNPass::processLoad(LoadInst *L) {
   reportLoadElim(L, AvailableValue, ORE);
   // Tell MDA to reexamine the reused pointer since we might have more
   // information after forwarding it.
+  doInstructionDeletion(L);
```
Move this above the comment about MDA (which is about the lines below).

madhur13490 replied:
Done

madhur13490 committed:
Address review comments

madhur13490 commented:
@nikic Removed the new includes from `GVN.h` and headers in `GVN.cpp` are untouched now.

nikic reviewed (approved):
LGTM
